### DISCUSSION TITLE
Plugins & modifiers from v4 to v5

### DISCUSSION THREAD

**Discussion by criocere** [smarty-php/smarty#1072] [Q&A]
Sorry but I don't really understand what to do with my plugins and modifiers to have them ok in v5? 
I have more than 50 plugins and same for modifiers, how can I register them all automatically as before?

In a folder **/plugins** I have for example the file **function.getPlantes.php**
```
function smarty_function_getPlantes($iParams, $iTemplate) 
{
....
}
```

Now I initialize smarty with

`$smarty->addPluginsDir(SMARTY.'/plugins');`

And now I understand that I have to set all the plugins one by one, right?

`$smarty->registerPlugin(Smarty\Smarty::PLUGIN_FUNCTION, 'getPlante', 'smarty_function_getPlante');`

Does smarty loads all the functions in the /plugins folder? 
The file name is function.getPlantes.php and the function name is smarty_function_getPlante, is that ok?

I don't really understand how to do my migration easily, the doc doesn't help very well.

Thanks for your help

**Comment by wisskid** [comment]
You could continue to use `addPluginsDir` and ignore the deprecation notice. It will still work.

Or you can create an Extension class like this:
```php
<?php

use Smarty\Extension\Base;

class MyExtension extends Base {

    public function getModifierCallback(string $modifierName) {

       switch ($modifierName) {
			case 'getPlantes': return [$this, 'smarty_function_getPlantes'];
       }

        return null;
    }


    public function smarty_function_getPlantes($iParams, $iTemplate) {
        // ....
    }
}
```

**Reply by wisskid** [reply]
See https://smarty-php.github.io/smarty/stable/api/extending/extensions/#registering-an-extension on how to register your new Extension class.

**Comment by criocere** [comment]
I have too many plugins and modifiers to do this extension one by one.
Is there a proper way to do that more generically as in V4? 
A kind of extension for V4 plugins and V4 modifiers, we can activate?
My problem is to find a solution before addPluginsDir will be removed.

Note: how to remove this deprecated message?

Thanks wisskid

**Reply by wisskid** [reply]
You cannot remove the deprecation error, but you can lower your [error reporting](https://www.php.net/manual/en/function.error-reporting.php) settings so PHP will not report or log them.

Writing a custom extension that reads an entire dir like addPluginsDir does shouldn't be too hard. You could write and release this as a separate composer package. Be aware though that when this is done for a website at runtime, this triggers a bunch of IO for each HTTP request. This may not be a problem for low-volume websites, but it's certainly not ideal for high traffic sites.

**Comment by criocere** [comment]
OK I spend my weekend to transfer all my functions and modifiers in 2 extensions... (as you explain). But impossible to have something working. I have always this fatal error

`PHP Fatal error:  Uncaught Error: Call to a member function handle() on null in /home/myUser/domains/myDomain/public_html/cache/smartyV5/templates/b2/d5/8b/b2d58b297cf536767115a6194e39c72a1393571c_0.file_plante2.tpl.php:85\nStack trace:\n#0 /home/myUser/domains/myDomain/public_html/inc/smartyV5/src/Runtime/InheritanceRuntime.php(240): Block_3601620426714d570a7a585_62135313->callBlock()\n#1 /home/myUser/domains/myDomain/public_html/inc/smartyV5/src/Runtime/InheritanceRuntime.php(178): Smarty\\Runtime\\InheritanceRuntime->callBlock()\n#2 /home/myUser/domains/myDomain/public_html/inc/smartyV5/src/Runtime/InheritanceRuntime.php(180): Smarty\\Runtime\\InheritanceRuntime->processBlock()\n#3 /home/myUser/domains/myDomain/public_html/inc/smartyV5/src/Runtime/InheritanceRuntime.php(150): Smarty\\Runtime\\InheritanceRuntime->processBlock()\n#4 /home/myUser/domains/myDomain/public_html/cache/smartyV5/templates/5d/5e/b8/5d5eb89b1cd14bd6f85af57ca617c40e54c84498_0.file...'`

This kind of message doesn't help to understand what's wrong. How to find the problem? How to debug?

**Reply by wisskid** [reply]
Can you show the compiled template that triggers this error?

**Comment by criocere** [comment]
Hereafter my code, certainly something I don't understand

**Extension for functions**

```
<?php
use Smarty\Extension\Base;
use Smarty\Exception;

class SmartyFunctionsExtension extends Base {

    public function getModifierCallback(string $name) {
        switch ($name) {
            case 'getActualitesArchives': 
            case 'getActualitesIndex': 
            //etc...
                return [$this, 'smarty_function_'.$name];
        }
        
        return null;
    }

    public function smarty_function_getActualitesArchives($iParams, $iTemplate) {  
         // my v4 code which works
    }

    // etc...
}
?>
```
       
**Extension for modifiers**
 
```
<?php
use Smarty\Extension\Base;
use Smarty\Exception;

class SmartyModifiersExtension extends Base {

    public function getModifierCallback(string $name) {
        switch($name) {
            case 'add':
            case 'addImageCopy':
            // etc...
                return [$this, 'smarty_modifier_'.$name];
        }
       
        return null;
    }

    public function smarty_modifier_add($number, $alteration = 1)    {
        return $number += $alteration;
    }

    // etc ....
}
?>
```

How I add those extensions in smarty

```
$smarty->addExtension(new SmartyFunctionsExtension());
$smarty->addExtension(new SmartyModifiersExtension());
$smarty->addExtension(new SmartyPHPExtension());    // this extension is OK
```

**Reply by wisskid** [reply]
Seems fine to me!

**Comment by criocere** [comment]
If I replace the extension with a loop for my functions with

`$smarty->registerPlugin(Smarty\Smarty::PLUGIN_FUNCTION, $f, 'smarty_function_'.$f);`

and for my modifiers with

`$smarty->registerPlugin(Smarty\Smarty::PLUGIN_MODIFIER, $m, 'smarty_modifier_'.$m);`

It works!!!! The list of modifiers and functions is the same.

Is there a way to have more traces to understand this error "Call to a member function handle() on null...."?

My templates are complex with inheritances and a lots of modifiers and functions. I cannot really simplify them to understand the problem

**Reply by wisskid** [reply]
The trace seems fine, just show us `/home/myUser/domains/myDomain/public_html/cache/smartyV5/templates/b2/d5/8b/b2d58b297cf536767115a6194e39c72a1393571c_0.file_plante2.tpl.php`, or at least line 85 and a few lines surrounding it.

**Reply by criocere** [reply]
I have those lines

```
    <?php if ((null !== ($_GET['debug'] ?? null))) {
$_smarty_tpl->assign('force', TRUE, false, NULL);
} else {
$_smarty_tpl->assign('force', FALSE, false, NULL);
}?>
    <?php echo $_smarty_tpl->getSmarty()->getFunctionHandler('getPlante')->handle(array('NomCommun'=>$_smarty_tpl->getValue('NomCommun'),'Photos'=>$_smarty_tpl->getValue('Image'),'PhotosTitres'=>(($tmp = $_smarty_tpl->getValue('ImageTitre') ?? null)===null||$tmp==='' ? $_smarty_tpl->getValue('Titre') ?? null : $tmp),'FamillesID'=>$_smarty_tpl->getValue('FamillesID'),'PagesIndexID'=>$_smarty_tpl->getValue('PagesIndexID'),'Genre'=>$_smarty_tpl->getValue('Genre'),'PlantesID'=>$_smarty_tpl->getValue('ID'),'RubriquesID'=>(($tmp = $_smarty_tpl->getValue('RubriquesID') ?? null)===null||$tmp==='' ? null ?? null : $tmp)), $_smarty_tpl);?>

    <link rel="preload" as="image" type="image/webp" media="(max-width: 599px)" href="<?php echo $_smarty_tpl->getSmarty()->getModifierCallback('resizeImage')($_smarty_tpl->getValue('tabPhotos')[0]['Photo'],420,315,((($tmp = $_smarty_tpl->getValue('tabPhotos')[0]['Auteur'] ?? null)===null||$tmp==='' ? 'www.auJardin.info' ?? null : $tmp)),$_smarty_tpl->getValue('force'),false,'webp');?>
">
    <link rel="preload" as="image" type="image/webp" media="(min-width: 600px)" href="<?php echo $_smarty_tpl->getSmarty()->getModifierCallback('resizeImage')($_smarty_tpl->getValue('tabPhotos')[0]['Photo'],600,450,((($tmp = $_smarty_tpl->getValue('tabPhotos')[0]['Auteur'] ?? null)===null||$tmp==='' ? 'www.auJardin.info' ?? null : $tmp)),$_smarty_tpl->getValue('force'),false,'webp');?>
">          
```

line 85 is this one

`   <?php echo $_smarty_tpl->getSmarty()->getFunctionHandler('getPlante')->handle(array('NomCommun'=>$_smarty_tpl->getValue('NomCommun'),'Photos'=>$_smarty_tpl->getValue('Image'),'PhotosTitres'=>(($tmp = $_smarty_tpl->getValue('ImageTitre') ?? null)===null||$tmp==='' ? $_smarty_tpl->getValue('Titre') ?? null : $tmp),'FamillesID'=>$_smarty_tpl->getValue('FamillesID'),'PagesIndexID'=>$_smarty_tpl->getValue('PagesIndexID'),'Genre'=>$_smarty_tpl->getValue('Genre'),'PlantesID'=>$_smarty_tpl->getValue('ID'),'RubriquesID'=>(($tmp = $_smarty_tpl->getValue('RubriquesID') ?? null)===null||$tmp==='' ? null ?? null : $tmp)), $_smarty_tpl);?>`

Thanks for your help

**Reply by wisskid** [reply]
In this case, you shouldn't have implemented it as a modifier callback, but as a FunctionHandler. See `\Smarty\Extension\ExtensionInterface::getFunctionHandler` and `\Smarty\FunctionHandler\FunctionHandlerInterface`.

**Reply by criocere** [reply]
Thanks for your answer but I think that the doc is not helpful on this migration of functions

**Reply by wisskid** [reply]
https://smarty-php.github.io/smarty/5.x/api/extending/extensions/#writing-your-own-extension has a couple of useful examples.

**Reply by criocere** [reply]
I wrote a kind of adapter for V4 plugins (function & modifier).
I post it hoping that this can help someone.
Maybe you can suggest a kind of solution like this in the doc to help migrating. Personally, I spent time to understand what to do and how to use extension with my v4 plugins to migrate in V5.
And thanks to your help

My custom extension

```
use Smarty\Extension\Base;
use Smarty\Exception;

include WEBROOT_INC.'/SmartyCustomFunction.class.php';

class SmartyExtension extends Base {

    private $modifiers = [];
    private $functionHandlers = [];

    public function getModifierCallback(string $name) 
    {
        // php modifiers
        $phpModifiers = ['urldecode', 'str_word_count', ...];
        if (\in_array($name, $phpModifiers)) return $name;

        // custom modifiers
        $customModifiers = ['modifier1', 'modifier2', ...];
        if (\in_array($name, $customModifiers)) {
            $modifierName = 'smarty_modifier_'.$name;
            if (!isset($this->modifiers[$name])) {
                $this->modifiers[$name] = $modifierName;
                include SMARTY.'/plugins/modifier.'.$name.'.php';
            }
            return $modifierName;
        }
        
        return null;
    }

    public function getFunctionHandler(string $name): ?\Smarty\FunctionHandler\FunctionHandlerInterface 
    {
        if (isset($this->functionHandlers[$name])) return $this->functionHandlers[$name];

        $customFunctions =['tag1', 'tag2', ...];
        if (\in_array($name, $customFunctions)) {
            $function = new \Smarty\FunctionHandler\SmartyCustomFunction($name);
            $this->functionHandlers[$name] = $function;
            return $function;
        }
        
        return null;
    }
}
```

My custom function handler
```
namespace Smarty\FunctionHandler;

use Smarty\Template;

class SmartyCustomFunction extends Base
{
    private string $functionName;

    public function __construct(string $functionName)
    {
        $this->functionName = $functionName;
        include SMARTY.'/plugins/function.'.$this->functionName.'.php';
    }

    public function handle($params, Template $template)
    {
        return \call_user_func('smarty_function_'.$this->functionName, $params, $template);
    }
}
```

Everything seems working
